### PR TITLE
Feature: Add akismet support to v3 forms

### PR DESCRIPTION
**Pull request by kjohnson**
Resolves [GIVE-1050]

## Description

This PR adds support for Akismet spam checks to the v3 donation forms.

## Affects

Adds the `givewp_donate_form_data_validated` action to `DonateRoute` where a hook can disrupt the donation process by throwing an `Exception`.

### Visuals

| Donation Flagged as SPAM |
| --- |
| image |

| SPAM Log |
| --- |
| image |

## Testing Instructions

- Install and activate the Akismet plugin
- Register a personal, free Akismet license
- Configure Akismet with the API key
- Enable "Akismet SPAM Protected" in the Give settings
- Donate with the email akismet-guaranteed-spam@example.com

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

[GIVE-1050]: https://stellarwp.atlassian.net/browse/GIVE-1050?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by kjohnson** [comment]
Build for QA, see https://github.com/impress-org/givewp/actions/runs/10220925772.

**Review by glaubersilva — CHANGES_REQUESTED** [review]
@kjohnson Great work! I loved the unit tests cause they always helped me better understand everything. 😊 I just left a few comments about minor changes.

File: src/DonationForms/Routes/DonateRoute.php [hunk 1]
```diff
@@ -55,6 +56,17 @@ public function __invoke(array $request)
...
+             * The donation flow can be interrupted by throwing an Exception.
+             *
+             * @unreleased
+             *
+             * @param DonateControllerData $data
+             */
+            do_action('givewp_donate_form_data_validated', $data);
+
             $this->donateController->donate($data, $data->getGateway());
         } catch (DonationFormFieldErrorsException $exception) {
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply src/DonationForms/Routes/DonateRoute.php:71]: I had considered that, but I didn't want the route error message to leak to a potential spammer. The donate controller routing suppresses the details when the error type is "unknown", which is intentional here.

In this case the exception error message is clear and is only available to the site admin (in the log).

File: src/DonationSpam/Akismet/DataTransferObjects/CommentCheckArgs.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    public $referrer;
+    public $comment_type;
+    public $comment_content;
+    public $comment_author;
+    public $comment_author_email;
+
+    /**
+     * @unreleased
+     */
+    public static function make(DonateControllerData $data)
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply src/DonationSpam/Akismet/DataTransferObjects/CommentCheckArgs.php]: Resolved by [c0e49e6](https://github.com/impress-org/givewp/pull/7474/commits/c0e49e6f437f9f17c5263e5c579f06b058a7766c).

[the remark this answers is not in the document]
Comment (kjohnson) [reply src/DonationSpam/Akismet/DataTransferObjects/CommentCheckArgs.php]: Resolved by [c0e49e6](https://github.com/impress-org/givewp/pull/7474/commits/c0e49e6f437f9f17c5263e5c579f06b058a7766c).

File: src/DonationSpam/Akismet/DataTransferObjects/CommentCheckArgs.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        }
+
+        return $self;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function toHttpQuery()
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply src/DonationSpam/Akismet/DataTransferObjects/CommentCheckArgs.php]: Resolved by [c0e49e6](https://github.com/impress-org/givewp/pull/7474/commits/c0e49e6f437f9f17c5263e5c579f06b058a7766c).

File: src/DonationSpam/ServiceProvider.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * @unreleased
+     * @return bool
+     */
+    public function isAkismetEnabledAndConfigured(): bool
+    {
+        return
+            give_is_setting_enabled( give_get_option( 'akismet_spam_protection' ) )
+            && give_check_akismet_key();
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply src/DonationSpam/ServiceProvider.php:45]: That would mean that the setting's configuration would be referencing itself.

```diff
[
    'name'    => __( 'Akismet SPAM Protection', 'give' ),
    'id'      => 'akismet_spam_protection*,
-   'default' => ( give_check_akismet_key() ) ? 'enabled' : 'disabled',
+   'default' => (give_is_setting_enabled(give_get_option('akismet_spam_protection'))) ? 'enabled' : 'disabled',
],
```

[the remark this answers is not in the document]
Comment (kjohnson) [reply src/DonationSpam/ServiceProvider.php:45]: If we call `give_check_akismet_key()` first, then we can assume a default value of `enabled` for the `akismet_spam_protection` option.

```php
public function isAkismetEnabledAndConfigured(): bool
{
    return
        give_check_akismet_key()
        && give_is_setting_enabled(
            give_get_option( 'akismet_spam_protection', 'enabled')
        );
}
```

File: tests/Unit/DonationSpam/Akismet/Actions/ValidateDonationTest.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+use Give\DonationForms\DataTransferObjects\DonateControllerData;
+use Give\DonationSpam\Akismet\Actions\ValidateDonation;
+use Give\DonationSpam\Akismet\API;
+use Give\DonationSpam\EmailAddressWhiteList;
+use Give\DonationSpam\Exceptions\SpamDonationException;
+use Give\Tests\TestCase;
+use PHPUnit_Framework_MockObject_MockBuilder;
+use PHPUnit_Framework_MockObject_MockObject;
+
+final class ValidateDonationTest extends TestCase
```
[the remark this answers is not in the document]
Comment (kjohnson) [reply tests/Unit/DonationSpam/Akismet/Actions/ValidateDonationTest.php:17]: Resolved, see [c6db18e](https://github.com/impress-org/givewp/pull/7474/commits/c6db18e17d72fe7220abd6feca628cf8855e1dc2).

File: src/DonationForms/Routes/DonateRoute.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/DonationForms/Routes/DonateRoute.php:71]: Got it. Thanks for the clarification!

File: src/DonationSpam/ServiceProvider.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/DonationSpam/ServiceProvider.php]: Oh yeah, it makes more sense. Thanks for the changes!

**Review by glaubersilva — APPROVED** [review]
@kjohnson Looks good to me, ready for QA. Great work!

**Review by rickalday — APPROVED** [review]
Passed manual QA tests.

**Pull request merged**